caps-spark/src/main/scala/org/opencypher/caps/impl/spark/convert/rowToCypherMap.scala

```
@@ -32,62 +33,72 @@ final case class rowToCypherMap(header: RecordHeader) extends (Row => CypherMap)
     CypherMap(values: _*)
   }
 
+  // TODO: Validate all column types. At the moment null values are cast to the expected type...
   private def constructValue(row: Row, field: Var): CypherValue = {
     field.cypherType match {
       case _: CTNode =>
-        val (id, labels, properties) = collectNode(row, field)
-        CAPSNode(id, labels.toSet, properties)
+        collectNode(row, field)
 
       case _: CTRelationship =>
-        val (id, source, target, typ, properties) = collectRel(row, field)
-        CAPSRelationship(id, source, target, typ, properties)
+        collectRel(row, field)
 
       case _ =>
         val raw = row.getAs[Any](SparkColumnName.of(header.slotFor(field)))
         CypherValue(raw)
     }
   }
 
-  private def collectNode(row: Row, field: Var): (Long, Seq[String], Map[String, CypherValue]) = {
-    val id = row.getAs[Long](SparkColumnName.of(header.slotFor(field)))
-    val labels = header
-      .labelSlots(field)
-      .mapValues { s =>
-        row.getAs[Boolean](SparkColumnName.of(s))
-      }
-      .collect {
-        case (h, b) if b =>
-          h.label.name
-      }
-      .toSeq
-    val properties = header
-      .propertySlots(field)
-      .mapValues { s =>
-        CypherValue(row.getAs[Any](SparkColumnName.of(s)))
-      }
-      .collect {
-        case (p, v) if !v.isNull =>
-          p.key.name -> v
-      }
+  private def collectNode(row: Row, field: Var): CypherValue = {
+    val idValue = row.getAs[Any](SparkColumnName.of(header.slotFor(field)))
+    idValue match {
+      case null       => CypherNull
```

This is invalid; the id can never be `null`. We should probably throw an exception, but I think we already set these kinds of columns as non-nullable in Spark.

---

It can, if we do an OptionalMatch

---

Ah, you're right!